### PR TITLE
HDDS-10477. Make Rocksdb tools native lib compatible with all chipset with the same arch

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?
While building rocksdb tools lib, makefile of rocksdb adds native optmization gcc flags when compiling rocksdb code. This makes the built code incompatible with other chipsets, where it fails with illegal instruction when loading the native lib on ozone manager jvm.

Setting PORTABLE=1 while building rocksdb tools will fix this particular issue, and also adding -O0 flag while compiling jni on ozone will prevent the same on ozone jnilib side.
https://github.com/facebook/rocksdb/blob/v7.7.3/INSTALL.md#:~:text=By%20default%20the,make%20static_lib.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10477

## How was this patch tested?
Successful build and checking the native lib manually on various instance

### CODE REVIEW

**Review by prashantpogde — APPROVED** [review]
Good catch. LGTM.

**Review by swagle — APPROVED** [review]
LGTM

File: hadoop-hdds/rocks-native/src/CMakeLists.txt [hunk 1]
```diff
@@ -22,8 +22,8 @@
 
 cmake_minimum_required(VERSION 2.8)
 add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
```
Comment (kerneltime) [inline hadoop-hdds/rocks-native/src/CMakeLists.txt:25]: Why no optimization at all? This is different that specifying the instruction sets. Can you look into `add_compile_options(-msse2) ` ? I am not sure no optimization == chipset instructions to be used.

**Comment by prashantpogde** [comment]
should we force USE_SSE=0 or -DFORCE_SSE42=OFF ?

**Comment by swamirishi** [comment]
> should we force USE_SSE=0 or -DFORCE_SSE42=OFF ?
This is off by default
Makefile doesn’t set USE_SSE by default. In the code the it checks if it has been set https://github.com/facebook/rocksdb/blob/eb9a80fe1f18017b4d7f4084e8f2554f12234822/build_tools/build_detect_platform#L663

File: hadoop-hdds/rocks-native/src/CMakeLists.txt (hunk 1 above)
Comment (swamirishi) [reply hadoop-hdds/rocks-native/src/CMakeLists.txt:25]: I am not really aware of this, but let me check this out and get back on this.

**Comment by kerneltime** [comment]
SSE2 is pretty common based on the [documentation in that makefile ](https://github.com/facebook/rocksdb/blob/eb9a80fe1f18017b4d7f4084e8f2554f12234822/build_tools/build_detect_platform#L713C1-L715C23)
```  # SSE4.2 available since nehalem, ca. 2008-2010
  # Includes POPCNT for BitsSetToOne, BitParity
  TRY_SSE42="-msse4.2"
```

**Comment by kerneltime** [comment]
@swamirishi the main issue is due to the use of AVX related instructions, we can skip those explicitly and we can target an instruction set that is more common.

File: hadoop-hdds/rocks-native/pom.xml [hunk 2]
```diff
@@ -235,6 +235,7 @@
                       <arg line="${project.build.directory}/rocksdb/rocksdb-${rocksdb.version}"/>
                     </exec>
                     <exec executable="make" dir="${project.build.directory}/rocksdb/rocksdb-${rocksdb.version}" failonerror="true">
+                      <arg line="PORTABLE=1"/>
```
Comment (kerneltime) [inline hadoop-hdds/rocks-native/pom.xml:238]: RocksDB does not use PORTABLE=0 so there is still some information missing as to what is the build setup.
For now pushing this in makes sense as we are seeing crashes for missing AVX4 but we still need to continue looking into what is the lowest common set of instruction set the build should support or at least quantify the performance impact. @swamirishi can you open a follow up Jira? Also, how was this change tested?

**Comment by kerneltime** [comment]
@swamirishi we can merge this in once you specify the testing done for this.

**Comment by swamirishi** [comment]
> @swamirishi we can merge this in once you specify the testing done for this.

I tested out this particular patch by building the code in an environment which supports avx2. I built the shared library and tried to load the library on another machine which doesn't support this instruction set without the flag PORTABLE=1 and with the flag PORTABLE=1. 
The library load failed with the error in the first run:
`#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007f995a5076b5, pid=110294, tid=0x00007f997fa59700
#
# JRE version: OpenJDK Runtime Environment (8.0_232-b09) (build 1.8.0_232-b09)
# Java VM: OpenJDK 64-Bit Server VM (25.232-b09 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libozone_rocksdb_tools.so+0x2276b5]  __static_initialization_and_destruction_0(int, int) [clone .constprop.534]+0x2445
#
# Core dump written. Default location: /root/core or core.110294
#
# An error report file with more information is saved as:
# /root/hs_err_pid110294.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Aborted (core dumped)
`
The second time the load succeeded.

**Comment by smengcl** [comment]
> > @swamirishi we can merge this in once you specify the testing done for this.
> 
> I tested out this particular patch by building the code in an environment which supports avx2. I built the shared library and tried to load the library on another machine which doesn't support this instruction set without the flag PORTABLE=1 and with the flag PORTABLE=1. The library load failed with the error in the first run:
> 
> Screen Shot 2024-03-20 at 15 15 19
> The second time the load succeeded.

Yup fyi `PORTABLE` flag is used here in `CMakeLists.txt` to disable avx/avx2 optimizations here:

https://github.com/facebook/rocksdb/blob/v7.7.3/CMakeLists.txt#L252-L277

File: hadoop-hdds/rocks-native/pom.xml (hunk 2 above)
Comment (smengcl) [reply hadoop-hdds/rocks-native/pom.xml:238]: The `.so` from the publicly distributed `rocksdbjni-7.7.3.jar` doesn't seem to be AVX or AVX2 optimized:

```bash
$ elfx86exts librocksdbjni-linux64.so 
File format and CPU architecture: Elf, X86_64
MODE64 (call)
CMOV (cmovbe)
SSE1 (movups)
SSE2 (movdqa)
BMI (tzcnt)
BMI2 (shlx)
Instruction set extensions used: BMI, BMI2, CMOV, MODE64, SSE1, SSE2
CPU Generation: Haswell
```

**Review by kerneltime — APPROVED** [review]

**Pull request merged**